### PR TITLE
In `pretty_print_type()`, print `async fn` futures' paths instead of spans.

### PR DESCRIPTION
**Pull request by kpreid**
This makes `-Zprint-type-sizes`'s output easier to read, because the name of an `async fn` is more immediately recognizable than its span. This change will also synergize with my other `-Zprint-type-sizes` PR #122922 which prints the type of child futures being awaited.

I also deleted the comment "FIXME(eddyb) should use `def_span`." because it appears to have already been fixed by commit 67727aa7c31a24ea73a91a9134c3653fae8209ab.

### CODE REVIEW

**Review by compiler-errors — COMMENTED** [review]
I don't mind this approach, but I think it needs some sort of delimiter to separate the "async fn body" part and the def path - `{async fn body operation}` is hard to parse over, e.g. `{async fn body @ operation}` or `{async fn body - operation}` something.

**Comment by kpreid** [comment]
> I don't mind this approach, but I think it needs some sort of delimiter to separate the "async fn body" part and the def path - `{async fn body operation}` is hard to parse over, e.g. `{async fn body @ operation}` or `{async fn body - operation}` something.

I was having that concern myself, but I couldn't think of a really good syntax. The "@" seems appropriate to keep only where it is currently used: before a span. How about `{async fn body of: operation}`?

**Comment by compiler-errors** [comment]
"of" seems a bit excessive. I think `:` is probably fine on its own. parens are fine i guess too

**Comment by compiler-errors** [comment]
Given this change is a bit more involved, I'm gonna cc @rust-lang/wg-diagnostics to see if anyone has thoughts. Given it's friday, I'm happy to r+ this on like monday or tuesday.

**Comment by kpreid** [comment]
A colon can have lots of meanings. My intuition about including the actual English word "of" is that new readers really don't know what all this punctuation and terminology means, and including "of" makes it much closer to immediately clear what we are talking about: this type is the "async fn body" _of_ the named function.

**Comment by kpreid** [comment]
I pushed a new version which prints like `{async fn body of name()}`, or in a more complex example from `tests/ui/async-await/future-sizes/async-awaiting-fut.stdout`,

```
std::mem::MaybeUninit<{async fn body of calls_fut<{async fn body of big_fut()}>()}>
```

I think that reads pretty well — the lack of turbofish isn't my fault. 😁

**Review by oli-obk — APPROVED** [review]

**Review by estebank — APPROVED** [review]

**Comment by compiler-errors** [comment]
ok we'll that's basically consensus

@bors r+

**Comment by kpreid** [comment]
Rebased and re-blessed on top of #122922 which uses this printing routine more often.

**Comment by workingjubilee** [comment]
@kpreid Needs rebase.

**Comment by kpreid** [comment]
Rebased.

**Comment by matthiaskrgr** [comment]
needs rebase I think https://github.com/rust-lang/rust/pull/123023#issuecomment-2017165606
@bors r-

**Comment by kpreid** [comment]
> needs rebase I think [#123023 (comment)](https://github.com/rust-lang/rust/pull/123023#issuecomment-2017165606)

Not actually a conflict (I think) but there was a file that hadn't gotten blessed before because I was having an unrelated problem actually finishing the run of `tests/mir-opt`. Sorry for the trouble. Fixed.
